### PR TITLE
Limit the number of deltas in RRDP notification.xml file. (#460)

### PR DESCRIPTION
**Pull request by timbru**

### CODE REVIEW

File: defaults/krill-pubd.conf [hunk 1]
```diff
@@ -122,6 +122,59 @@
...
+# this because:
+#  1) big notification files are expensive to serve and parse
+#  2) it's not just size, because of latency parsing many small deltas may
+#     be more time consuming than falling back to a delta
+#
+# So, the objective would be to optimize things such that the notification.xml files
+# remain relatively small, while still allowing the vast majority of Relying Party
+# implementations to use deltas, under normal operations.
+#
+# We have use the following strategy to prune excessive deltas:
```
Comment (ximon18) [inline defaults/krill-pubd.conf:148]: Remove "have" ?

File: defaults/krill-pubd.conf [hunk 2]
```diff
@@ -122,6 +122,59 @@
...
+#  - always keep 'retention_delta_files_min_nr' files
+#  - always keep 'retention_delta_files_min_seconds' files
+#  - beyond this:
+#    - never keep more than 'retention_delta_files_max_nr'
+#    - never keep older than 'retention_delta_files_max_seconds'
+#    - keep the others
+#
+# With these defaults:
+# retention_delta_files_min_nr = 5
+# dflt_retention_delta_files_min_seconds = 1200 (20 mins: twice the interval of commonly used RPs)
```
Comment (ximon18) [inline defaults/krill-pubd.conf:160]: Why do these settings have `dflt_` in the name? If you edit them does it change only a default but have no impact, or does it actually change behaviour and thus they should not be called defaults? Or are these just comments and the settings you can change are defined somewhere else?

File: defaults/krill-pubd.conf [hunk 3]
```diff
@@ -122,6 +122,59 @@
...
+# dflt_retention_delta_files_max_seconds = 7200 (2 hours: twice the interval of slowest RPs)
+# dflt_retention_delta_files_max_nr = 50
+
+# Time to keep any files still referenced by notification files updated up to X seconds ago. We
+# should not delete these files too eagerly or we would risk that RPs with an old notification file
+# try to retrieve them, without success.
+#
+# retention_old_notification_files_seconds = 600 (10 mins)
```
Comment (ximon18) [inline defaults/krill-pubd.conf:168]: Does the value format of this setting support "(10 mins)" - I assume not. I prefer it when commented out settings just work if the leading comment character is removed.

File: src/daemon/config.rs [hunk 4]
```diff
@@ -303,40 +303,57 @@ pub struct IssuanceTimingConfig {
 pub struct RepositoryRetentionConfig {
     #[serde(default = "RepositoryRetentionConfig::dflt_retention_old_notification_files_seconds")]
     pub retention_old_notification_files_seconds: i64,
-    #[serde(default = "RepositoryRetentionConfig::dflt_retention_delta_files_seconds")]
-    pub retention_delta_files_seconds: i64,
     #[serde(default = "RepositoryRetentionConfig::dflt_retention_delta_files_min_nr")]
     pub retention_delta_files_min_nr: usize,
```
Comment (ximon18) [inline src/daemon/config.rs:307]: These settings don't seem to be documented in `krill-pubd.conf`. They are only present with the `dflt_` prefix, not without it.

File: src/daemon/config.rs [hunk 5]
```diff
@@ -508,8 +525,10 @@ impl Config {
 
         let repository_retention = RepositoryRetentionConfig {
             retention_old_notification_files_seconds: 1,
-            retention_delta_files_seconds: 1,
+            retention_delta_files_min_seconds: 0,
             retention_delta_files_min_nr: 5,
+            retention_delta_files_max_seconds: 1,
```
Comment (ximon18) [inline src/daemon/config.rs:530]: max seconds 1? Is this correct?

File: src/pubd/repository.rs [hunk 6]
```diff
@@ -556,25 +556,41 @@ impl RrdpServer {
...
+        let max_nr = config.retention_delta_files_max_nr;
+        let max_secs = config.retention_delta_files_max_seconds;
 
         for delta in &self.deltas {
             size += delta.elements().size();
-            if size > snapshot_size || (keep > minimum && delta.older_than_seconds(retain_secs)) {
-                break;
+
+            if size > snapshot_size {
+                // never keep more than the size of the snapshot
```
Comment (ximon18) [inline src/pubd/repository.rs:582]: Given that size will only keep increasing as you iterate over the loop, maybe break out here?

File: src/pubd/repository.rs [hunk 7]
```diff
@@ -556,25 +556,41 @@ impl RrdpServer {
...
+            } else if keep < min_nr || delta.younger_than_seconds(min_secs) {
+                // always keep 'retention_delta_files_min_nr' files
+                // always keep 'retention_delta_files_min_seconds' file
+                keep += 1
+            } else if keep == max_nr || delta.older_than_seconds(max_secs) {
```
Comment (ximon18) [inline src/pubd/repository.rs:587]: Is self.deltas ordered by age? If so you break out of the loop here. If not, what happens if one delta is older sthan max seconds so is skipped but we go round the loop again and find the next delta is _younger than max seconds and so is kept, but now keep has incremented by 1 referring to keeping the last delta, not this one??? This looks correct but feels like there is potential for errors here.

File: defaults/krill-pubd.conf (hunk 1 above)
Comment (timbru) [reply defaults/krill-pubd.conf:148]: ack

File: defaults/krill-pubd.conf (hunk 2 above)
Comment (timbru) [reply defaults/krill-pubd.conf:160]: removed dflt_

File: defaults/krill-pubd.conf (hunk 3 above)
Comment (timbru) [reply defaults/krill-pubd.conf:168]: indeed, removed

File: src/daemon/config.rs (hunk 4 above)
Comment (timbru) [reply src/daemon/config.rs:307]: removed dflt_

File: src/pubd/repository.rs (hunk 6 above)
Comment (timbru) [reply src/pubd/repository.rs:582]: yep, now breaking

File: src/pubd/repository.rs (hunk 7 above)
Comment (timbru) [reply src/pubd/repository.rs:587]: yep, now breaking

File: src/daemon/config.rs (hunk 5 above)
Comment (timbru) [reply src/daemon/config.rs:530]: this is test config, so yes

Comment (ximon18) [reply src/daemon/config.rs:530]: Ah, oops :-)

**Review by ximon18 — APPROVED** [review]

**Pull request merged**